File: java/src/main/java/com/weizilla/adventofcode/utils/Day.java
package com.weizilla.adventofcode.utils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class Day {
    private static final Logger logger = LoggerFactory.getLogger(Day.class);
    protected final InputReader reader;
    private final int year;
    private final int day;
    private final Integer example;

    protected Day(Integer example) {
        RunDay runDay = Utils.getRunDay();
        this.year = runDay.year();
        this.day = runDay.day();
        this.example = example;
        reader = new InputReader(year, day, example);
    }

    public Object part1() {
        // not implemented yet!
        return null;
    }

    public Object part2() {
        return null;
    }

    public void printAlways(Object input) {
        printAlways("{}", input);
    }

    public void printAlways(String input, Object ... args) {
        logger.info(input, args);
    }

    public void print(Object input) {
        print("{}", input);
    }

    public void print(String input, Object ... args) {
        if (example != null) {
            printAlways(input, args);
        }
    }

    public int getYear() {
        return year;
    }

    public int getDay() {
        return day;
    }

    public Integer getExample() {
        return example;
    }
}

File: scala/src/main/scala/com/weizilla/adventofcode/utils/Day.scala
package com.weizilla.adventofcode
package utils

import org.slf4j.LoggerFactory

class Day(year: Int, day: Int, example: Integer) {
  private val logger = LoggerFactory.getLogger(this.getClass)
  protected val reader = new InputReader(year, day, example)

  def part1(): Any = { }

  def part2(): Any = { }

  def printAlways(input: Any, args: Any *): Unit = {
    if (args.isEmpty) {
      logger.info("{}", input)
    } else {
      logger.info(input.toString, args: _*)
    }
  }

  def print(input: Any, args: Any *): Unit = {
    if (example != null) {
      if (args.isEmpty) {
        printAlways("{}", input)
      } else {
        printAlways(input, args: _*)
      }
    }
  }

  def getYear: Int = {
    year
  }

  def getDay: Int = {
    day
  }
}
